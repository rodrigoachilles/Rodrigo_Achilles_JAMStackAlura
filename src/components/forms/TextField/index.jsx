import PropTypes from 'prop-types';
import React from 'react';
import Text from '../../foundation/Text';
import Input, { InputWrapper } from './styles';

export default function TextField({
  type,
  placeholder,
  name,
  onChange,
  value,
  error,
  isTouched,
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;
  
  return (
    <InputWrapper>
      <Input
        type={type}
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
      />
      
      {isFieldInvalid && (
        <Text
          variant="smallestException"
          color="error.main"
          role="alert"
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}

TextField.defaultProps = {
  type: 'text',
  error: '',
  isTouched: false,
};

TextField.propTypes = {
  type: PropTypes.string,
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
};
