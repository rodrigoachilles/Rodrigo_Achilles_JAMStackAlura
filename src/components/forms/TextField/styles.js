import styled from 'styled-components';
import Text from '../../foundation/Text';

export const InputWrapper = styled.div`
  margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  border: 1px solid ${({ theme }) => theme.colors.borders.main.color};
  padding: 12px 16px;
  outline: 0;
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph',
};

export default Input;
