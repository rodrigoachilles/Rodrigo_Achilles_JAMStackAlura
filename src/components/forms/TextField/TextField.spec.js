import user from '@testing-library/user-event';
import React from 'react';
import { render, screen } from '../../../infra/test/testUtils';
import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Rodrigo"
        onChange={() => {}}
        name="nome"
      />
      ,
    );
    // screen.debug();
    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn();
        render(
          <TextField
            placeholder="Nome"
            value=""
            onChange={onChangeMock}
            name="nome"
            isTouched
          />
          ,
        );

        const inputNome = screen.getByPlaceholderText(/nome/i);
        user.type(inputNome, 'Rodrigo Achilles');
        expect(onChangeMock).toHaveBeenCalledTimes(16);
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          placeholder="Email"
          value="rodrigo.achilles@"
          onChange={() => {}}
          name="email"
          isTouched
          error="O campo email é obrigatório"
        />
        ,
      );

      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(inputEmail).toHaveValue('rodrigo.achilles@');
      expect(screen.getByRole('alert')).toHaveTextContent('O campo email é obrigatório');
      expect(inputEmail).toMatchSnapshot();
    });
  });
});
