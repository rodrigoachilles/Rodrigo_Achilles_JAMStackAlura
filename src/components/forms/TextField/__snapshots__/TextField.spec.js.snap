// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-family: 'Fira Sans';
  font-size: 12px;
  font-weight: normal;
  line-height: 1;
}

.c1 {
  width: 100%;
  border: 1px solid #99C1CC;
  padding: 12px 16px;
  outline: 0;
}

@media only screen and (min-width:768px) {
  .c0 {
    font-size: 16px;
    font-weight: normal;
    line-height: 1;
  }
}

<input
  class="c0 c1"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Rodrigo"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-family: 'Fira Sans';
  font-size: 12px;
  font-weight: normal;
  line-height: 1;
}

.c1 {
  width: 100%;
  border: 1px solid #99C1CC;
  padding: 12px 16px;
  outline: 0;
}

@media only screen and (min-width:768px) {
  .c0 {
    font-size: 16px;
    font-weight: normal;
    line-height: 1;
  }
}

<input
  class="c0 c1"
  name="email"
  placeholder="Email"
  type="text"
  value="rodrigo.achilles@"
/>
`;
